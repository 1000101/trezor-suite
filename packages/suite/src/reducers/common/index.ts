import router from './router';
import suite from './suite';

export default {
    suite,
    router,
};
