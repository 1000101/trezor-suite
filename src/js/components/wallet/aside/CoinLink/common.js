import PropTypes from 'prop-types';

export const coinProp = {
    coin: PropTypes.shape({
        img: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
    }).isRequired,
};
