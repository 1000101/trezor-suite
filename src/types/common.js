/* @flow */

/* Common types used in both params and responses */

export type BlockchainSettings = {
    name: string,
    worker: string | Function,
    server: Array<string>,
    debug?: boolean,
};

/* Transaction */

export type TokenTransfer = {
    type: 'sent' | 'recv' | 'self' | 'unknown',
    name: string,
    symbol: string,
    address: string,
    decimals: number,
    amount: string,
    from?: string,
    to?: string,
};

export type Target = {
    addresses?: Array<string>,
    isAddress: boolean,
    amount?: string,
    coinbase?: string,
};

export type Transaction = {
    type: 'sent' | 'recv' | 'self' | 'unknown',

    txid: string,
    blockTime: ?number,
    blockHeight: ?number,
    blockHash: ?string,

    amount: string,
    fee: string,
    // total: string, // amount + total

    targets: Array<Target>,
    tokens: Array<TokenTransfer>,
};

/* Account */

export type Address = {
    address: string,
    path: string,
    transfers: number,
    // decimal: number,
    balance?: string,
    sent?: string,
    received?: string,
};

export type AccountAddresses = {
    change: Array<Address>,
    used: Array<Address>,
    unused: Array<Address>,
};

export type TokenInfo = {
    type: string, // token type: ERC20...
    address: string, // token address
    balance: string, // token balance
    name: string, // token name
    symbol: string, // token symbol
    decimals: number, //
    // transfers: number, // total transactions?
};

export type AccountInfo = {
    descriptor: string,
    balance: string,
    availableBalance: string,
    tokens?: Array<TokenInfo>, // ethereum tokens
    addresses?: AccountAddresses, // bitcoin addresses
    history: {
        total: number, // total transactions (unknown in ripple)
        tokens?: number, // tokens transactions
        unconfirmed: number, // unconfirmed transactions
        transactions?: Array<Transaction>, // list of transactions
        txids?: Array<string>, // not implemented
    },
    misc?: {
        // ETH
        nonce?: string,
        // XRP
        sequence?: number,
        reserve?: string,
    },
    page?: ?{
        // blockbook
        index: number,
        size: number,
        total: number,
    },
    marker?: ?{
        // ripple-lib
        ledger: number,
        seq: number,
    },
};

export type SubscriptionAccountInfo = {
    descriptor: string,
    addresses?: AccountAddresses, // bitcoin addresses
};
