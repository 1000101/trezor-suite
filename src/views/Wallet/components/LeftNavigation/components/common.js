/* @flow */
import * as TrezorConnectActions from 'actions/TrezorConnectActions';
import * as RouterActions from 'actions/RouterActions';
import { toggleDeviceDropdown } from 'actions/WalletActions';
import type { State } from 'flowtype';

export type StateProps = {
    connect: $ElementType<State, 'connect'>,
    accounts: $ElementType<State, 'accounts'>,
    router: $ElementType<State, 'router'>,
    deviceDropdownOpened: boolean,
    fiat: $ElementType<State, 'fiat'>,
    localStorage: $ElementType<State, 'localStorage'>,
    discovery: $ElementType<State, 'discovery'>,
    wallet: $ElementType<State, 'wallet'>,
    devices: $ElementType<State, 'devices'>,
    pending: $ElementType<State, 'pending'>,
}

export type DispatchProps = {
    toggleDeviceDropdown: typeof toggleDeviceDropdown,
    addAccount: typeof TrezorConnectActions.addAccount,
    acquireDevice: typeof TrezorConnectActions.acquire,
    forgetDevice: typeof TrezorConnectActions.forget,
    duplicateDevice: typeof TrezorConnectActions.duplicateDevice,
    gotoDeviceSettings: typeof RouterActions.gotoDeviceSettings,
    onSelectDevice: typeof RouterActions.selectDevice,
}

export type Props = StateProps & DispatchProps;