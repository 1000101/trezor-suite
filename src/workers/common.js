/* @flow */

import type { Response, BlockchainSettings, SubscriptionAccountInfo } from '../types';
import { MESSAGES, RESPONSES } from '../constants';
import { CustomError } from '../constants/errors';

declare function postMessage(data: Response): void;

let _settings: BlockchainSettings;
let _debugPrefix: string;
let _addresses: Array<string> = [];
let _accounts: Array<SubscriptionAccountInfo> = [];
const _subscription: { [key: string]: boolean } = {};

const removeEmpty = (obj: Object) => {
    Object.keys(obj).forEach(key => {
        if (obj[key] && typeof obj[key] === 'object') removeEmpty(obj[key]);
        else if (obj[key] === undefined) delete obj[key];
    });
    return obj;
};

export const setSettings = (s: BlockchainSettings): void => {
    _settings = s;
    _debugPrefix = `[Worker "${s.name}"]:`;
};

export const getSettings = (): BlockchainSettings => _settings;

export const debug = (...args: Array<any>): void => {
    if (_settings && _settings.debug) {
        if (args[0] === 'warn' || args[0] === 'error') {
            console[args[0]](_debugPrefix, ...args.slice(1));
        } else {
            console.log(_debugPrefix, ...args);
        }
    }
};

export const handshake = (): void => {
    postMessage({
        id: -1,
        type: MESSAGES.HANDSHAKE,
    });
};

export const errorHandler = ({ id, error }: { id: number, error: any }): void => {
    let code: string = 'blockchain_link/unknown';
    // let message: string = '';
    // this part is uncovered by tests but for some reason i've write it like this (ripple maybe)
    // TODO: remove it after
    // if (typeof error === 'string') {
    //     message = error;
    // } else if (typeof error === 'object') {
    //     const keys = Object.keys(error);
    //     if (keys.indexOf('name') >= 0) {
    //         message = error.name;
    //     } else {
    //         message = error.message;
    //     }
    //     if (error.code) {
    //         code = error.code;
    //     }
    // }
    const message: string = error.message;
    if (error.code) {
        code = error.code;
    }

    postMessage({
        id,
        type: RESPONSES.ERROR,
        payload: {
            code,
            message,
        },
    });
};

export const response = (data: Response): void => {
    postMessage(removeEmpty(data));
};

const validateAddresses = (addresses: Array<string>): Array<string> => {
    if (!Array.isArray(addresses)) throw new CustomError('invalid_param', '+addresses');
    const seen = {};
    return addresses.filter(a => {
        if (typeof a !== 'string') return false;
        return seen.hasOwnProperty(a) ? false : (seen[a] = true);
    });
};

export const addAddresses = (addresses: Array<string>): Array<string> => {
    const unique = validateAddresses(addresses).filter(a => _addresses.indexOf(a) < 0);
    _addresses = _addresses.concat(unique);
    return unique;
};

export const getAddresses = (): Array<string> => _addresses;

export const removeAddresses = (addresses: Array<string>): Array<string> => {
    const unique = validateAddresses(addresses);
    _addresses = _addresses.filter(a => unique.indexOf(a) < 0);
    return _addresses;
};

const validateAccounts = (
    accounts: Array<SubscriptionAccountInfo>
): Array<SubscriptionAccountInfo> => {
    if (!Array.isArray(accounts)) throw new CustomError('invalid_param', '+accounts');
    const seen = {};
    return accounts.filter(a => {
        if (a && typeof a === 'object' && typeof a.descriptor === 'string') {
            return seen.hasOwnProperty(a.descriptor) ? false : (seen[a.descriptor] = true);
        }
        return false;
    });
};

const getAccountAddresses = (account): Array<string> => {
    // if (typeof account !== 'object' && typeof account.descriptor !== 'string') return [];
    if (account.addresses) {
        return []
            .concat(account.addresses.change, account.addresses.used, account.addresses.unused)
            .map(a => a.address);
    }
    return [account.descriptor];
};

export const addAccounts = (
    accounts: Array<SubscriptionAccountInfo>
): Array<SubscriptionAccountInfo> => {
    const valid = validateAccounts(accounts);
    const others = _accounts.filter(a => !valid.find(b => b.descriptor === a.descriptor));
    _accounts = others.concat(valid);
    const addresses = _accounts.reduce((addr, acc) => {
        return addr.concat(getAccountAddresses(acc));
    }, []);
    addAddresses(addresses);
    return valid;
};

export const getAccount = (address: string): ?SubscriptionAccountInfo => {
    return _accounts.find(a => {
        if (a.descriptor === address) return true;
        // console.log("2. get acc", a)
        if (a.addresses) {
            const { change, used, unused } = a.addresses;
            if (change.find(ad => ad.address === address)) return true;
            if (used.find(ad => ad.address === address)) return true;
            if (unused.find(ad => ad.address === address)) return true;
        }
        return false;
    });
};

export const getAccounts = (): Array<SubscriptionAccountInfo> => _accounts;

export const removeAccounts = (
    accounts: Array<SubscriptionAccountInfo>
): Array<SubscriptionAccountInfo> => {
    const valid = validateAccounts(accounts);
    const accountsToRemove = _accounts.filter(a => valid.find(b => b.descriptor === a.descriptor));
    const addressesToRemove = accountsToRemove.reduce((addr, acc) => {
        return addr.concat(getAccountAddresses(acc));
    }, []);
    _accounts = _accounts.filter(a => accountsToRemove.indexOf(a) < 0);
    removeAddresses(addressesToRemove);
    return _accounts;
};

export const addSubscription = (type: string): void => {
    _subscription[type] = true;
};

export const getSubscription = (type: string): boolean => _subscription[type];

export const removeSubscription = (type: string): void => {
    delete _subscription[type];
};

export const clearSubscriptions = (): void => {
    Object.keys(_subscription).forEach(key => {
        delete _subscription[key];
    });
};

export const shuffleEndpoints = (a: Array<string>): Array<string> => {
    for (let i = a.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        // $FlowIssue
        [a[i], a[j]] = [a[j], a[i]];
    }
    return a;
};
