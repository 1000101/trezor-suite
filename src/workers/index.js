/* @flow */

// WebWorkers manager

/* $FlowIssue loader notation */
import RippleWorker from 'worker-loader?name=js/ripple-worker.[hash].js!./ripple/index.js';
import { create as createDeferred } from '../utils/deffered';

import { MESSAGES, RESPONSES } from '../constants';
import type { Message, Response, Deferred } from '../types';

const sockets: Array<Socket> = [];

export const send = async (message: Message) => {
    if (!message.type) throw new Error('Message type not found');
    if (!message.network) throw new Error('Message network not found');

    let socket = sockets.find(s => s.network === message.network);
    if (!socket) {
        socket = new Socket(message.network);
        await socket.init();
        sockets.push(socket);
    }
    return await socket.send(message);
}

class Socket {
    messageId: number = 0;
    network: string;
    worker: Worker;
    deferred: Array<Deferred<any>> = [];
    notificationHandler: (event: any) => void = null;

    constructor(network: string) {
        this.network = network;
        this.worker = new RippleWorker();
        this.worker.onmessage = this.onMessage.bind(this);
        this.worker.onerror = this.onError.bind(this);
    }

    async init() {
        return await this.send({ type: MESSAGES.INIT });
    }

    async send(message: Message) {
        if (message.type === MESSAGES.SUBSCRIBE && message.notificationHandler) {
            this.notificationHandler = message.notificationHandler;
            // handler must not be passed to the worker
            message.notificationHandler = undefined;
        }
        const dfd = createDeferred(this.messageId);
        this.deferred.push(dfd);
        this.worker.postMessage({ id: this.messageId, ...message });
        this.messageId++;
        return await dfd.promise;
    }

    onMessage(event: { data: Response }) {
        if (!event.data) return;
        if (event.data.type == RESPONSES.NOTIFICATION) {
            if (this.notificationHandler) {
                this.notificationHandler(event.data.event);
            }
        } else {
            const dfd = this.deferred.find(d => d.id === event.data.id);
            if (!dfd) return;
            dfd.resolve(event.data);
        }
    }

    onError() {

    }

    close() {

    }
}