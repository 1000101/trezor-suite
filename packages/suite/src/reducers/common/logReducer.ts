/* @flow */

import { LOG } from '@suite-actions/constants';

import { Action } from '@suite/types';

export interface LogEntry {
    time: number;
    type: string;
    message: any;
}

export interface State {
    opened: boolean;
    entries: LogEntry[];
    copied: boolean;
}

export const initialState: State = {
    opened: false,
    entries: [],
    copied: false,
};

export default (state: State = initialState, action: Action): State => {
    switch (action.type) {
        case LOG.OPEN:
            return {
                ...state,
                opened: true,
            };

        case LOG.CLOSE:
            return {
                ...state,
                opened: false,
                copied: false,
            };

        case LOG.ADD:
            return {
                ...state,
                entries: state.entries.concat([action.payload]),
            };

        case LOG.COPY_SUCCESS:
            return {
                ...state,
                copied: true,
            };

        case LOG.COPY_RESET:
            return {
                ...state,
                copied: false,
            };

        default:
            return state;
    }
};
