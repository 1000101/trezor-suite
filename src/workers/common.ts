import { Response, BlockchainSettings, SubscriptionAccountInfo } from '../types';
import { MESSAGES, RESPONSES } from '../constants';
import { CustomError } from '../constants/errors';

declare function postMessage(data: Response): void;

let _settings: BlockchainSettings;
let _debugPrefix: string;
let _addresses: string[] = [];
let _accounts: SubscriptionAccountInfo[] = [];
const _subscription: { [key: string]: boolean } = {};

const removeEmpty = (obj: Response) => {
    Object.keys(obj).forEach(key => {
        if (obj[key] && typeof obj[key] === 'object') removeEmpty(obj[key]);
        else if (obj[key] === undefined) delete obj[key];
    });
    return obj;
};

export const setSettings = (s: BlockchainSettings): void => {
    _settings = s;
    _debugPrefix = `[Worker "${s.name}"]:`;
};

export const getSettings = (): BlockchainSettings => _settings;

export const debug = (...args: any[]): void => {
    if (_settings && _settings.debug) {
        if (args[0] === 'warn' || args[0] === 'error') {
            console[args[0]](_debugPrefix, ...args.slice(1));
        } else {
            console.log(_debugPrefix, ...args);
        }
    }
};

export const handshake = (): void => {
    postMessage({
        id: -1,
        type: MESSAGES.HANDSHAKE,
    });
};

export const errorHandler = ({ id, error }: { id: number, error: any }): void => {
    let code: string = 'blockchain_link/unknown';
    // let message: string = '';
    // this part is uncovered by tests but for some reason i've write it like this (ripple maybe)
    // TODO: remove it after
    // if (typeof error === 'string') {
    //     message = error;
    // } else if (typeof error === 'object') {
    //     const keys = Object.keys(error);
    //     if (keys.indexOf('name') >= 0) {
    //         message = error.name;
    //     } else {
    //         message = error.message;
    //     }
    //     if (error.code) {
    //         code = error.code;
    //     }
    // }
    const message: string = error.message;
    if (error.code) {
        code = error.code;
    }

    postMessage({
        id,
        type: RESPONSES.ERROR,
        payload: {
            code,
            message,
        },
    });
};

export const response = (data: Response) => {
    postMessage(removeEmpty(data));
};

const validateAddresses = (addresses: string[]) => {
    if (!Array.isArray(addresses)) throw new CustomError('invalid_param', '+addresses');
    const seen = {};
    return addresses.filter(a => {
        if (typeof a !== 'string') return false;
        return seen.hasOwnProperty(a) ? false : (seen[a] = true);
    });
};

export const addAddresses = (addresses: string[]) => {
    const unique = validateAddresses(addresses).filter(a => _addresses.indexOf(a) < 0);
    _addresses = _addresses.concat(unique);
    return unique;
};

export const getAddresses = () => _addresses;

export const removeAddresses = (addresses: string[]) => {
    const unique = validateAddresses(addresses);
    _addresses = _addresses.filter(a => unique.indexOf(a) < 0);
    return _addresses;
};

const validateAccounts = (
    accounts: SubscriptionAccountInfo[]
): SubscriptionAccountInfo[] => {
    if (!Array.isArray(accounts)) throw new CustomError('invalid_param', '+accounts');
    const seen = {};
    return accounts.filter(a => {
        if (a && typeof a === 'object' && typeof a.descriptor === 'string') {
            return seen.hasOwnProperty(a.descriptor) ? false : (seen[a.descriptor] = true);
        }
        return false;
    });
};

const getAccountAddresses = (account: SubscriptionAccountInfo) => {
    if (account.addresses) {
        const { change, used, unused } = account.addresses;
        return change.concat(used, unused).map(a => a.address);
    }
    return [account.descriptor];
};

export const addAccounts = (
    accounts: SubscriptionAccountInfo[]
): SubscriptionAccountInfo[] => {
    const valid = validateAccounts(accounts);
    const others = _accounts.filter(a => !valid.find(b => b.descriptor === a.descriptor));
    _accounts = others.concat(valid);
    const addresses = _accounts.reduce((addr, acc) => {
        return addr.concat(getAccountAddresses(acc));
    }, [] as string[]);
    addAddresses(addresses);
    return valid;
};

export const getAccount = (address: string): SubscriptionAccountInfo | undefined => {
    return _accounts.find(a => {
        if (a.descriptor === address) return true;
        if (a.addresses) {
            const { change, used, unused } = a.addresses;
            if (change.find(ad => ad.address === address)) return true;
            if (used.find(ad => ad.address === address)) return true;
            if (unused.find(ad => ad.address === address)) return true;
        }
        return false;
    });
};

export const getAccounts = () => _accounts;

export const removeAccounts = (
    accounts: SubscriptionAccountInfo[]
): SubscriptionAccountInfo[] => {
    const valid = validateAccounts(accounts);
    const accountsToRemove = _accounts.filter(a => valid.find(b => b.descriptor === a.descriptor));
    const addressesToRemove = accountsToRemove.reduce((addr, acc) => {
        return addr.concat(getAccountAddresses(acc));
    }, [] as string[]);
    _accounts = _accounts.filter(a => accountsToRemove.indexOf(a) < 0);
    removeAddresses(addressesToRemove);
    return _accounts;
};

export const addSubscription = (type: string): void => {
    _subscription[type] = true;
};

export const getSubscription = (type: string): boolean => _subscription[type];

export const removeSubscription = (type: string): void => {
    delete _subscription[type];
};

export const clearSubscriptions = (): void => {
    Object.keys(_subscription).forEach(key => {
        delete _subscription[key];
    });
};

export const shuffleEndpoints = (a: string[]) => {
    for (let i = a.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [a[i], a[j]] = [a[j], a[i]];
    }
    return a;
};
