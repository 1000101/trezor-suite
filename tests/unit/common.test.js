import * as common from '../../src/workers/common';

describe('Common:', () => {
    it('add unique address', () => {
        const data = [
            'not-an-array',
            {},
            new Error('error'),
            ['A', 'A', 'A', { invalid: 'value' }],
            ['A', 'B', 'A', false],
            ['A', 'B', 'C', 'A', 'D'],
        ];
        data.forEach((item) => {
            expect(common.addAddresses(item)).toMatchSnapshot();
        });
    });

    it('remove address from the list', () => {
        const data = [
            'not-an-array',
            ['A', undefined],
            ['B', 'B', 'B'],
            ['C', {}, 'D'],
        ];
        data.forEach((item) => {
            expect(common.removeAddresses(item)).toMatchSnapshot();
        });
    });
});