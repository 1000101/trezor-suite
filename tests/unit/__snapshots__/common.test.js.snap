// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common: add unique address 1`] = `Array []`;

exports[`Common: add unique address 2`] = `Array []`;

exports[`Common: add unique address 3`] = `Array []`;

exports[`Common: add unique address 4`] = `
Array [
  "A",
]
`;

exports[`Common: add unique address 5`] = `
Array [
  "B",
]
`;

exports[`Common: add unique address 6`] = `
Array [
  "C",
  "D",
]
`;

exports[`Common: remove address from the list 1`] = `
Array [
  "A",
  "B",
  "C",
  "D",
]
`;

exports[`Common: remove address from the list 2`] = `
Array [
  "B",
  "C",
  "D",
]
`;

exports[`Common: remove address from the list 3`] = `
Array [
  "C",
  "D",
]
`;

exports[`Common: remove address from the list 4`] = `Array []`;
